docs/source/development/eep-02-typing.md

```diff
@@ -0,0 +1,1474 @@
+future. A rough prototype of `BenchmarkScaler` looks as follows:
+
+```python
+class BenchmarkScaler(ABC):
+    @abstractmethod
+    def scale(self, params: NDArray) -> NDArray:
+        pass
+
+    @abstractmethod
+    def unscale(self, params: NDArray) -> NDArray:
+        pass
+```
+
+##### Representing benchmark problems
+
+Instead of the fixed-field dictionary we will have a dataclass with corresponding
+fields. This would roughly look as follows:
+
+```python
+@dataclass
+class BenchmarkProblem:
+    fun: Callable[[NDArray], FunctionValue]
+    start_x: NDArray
+    solution_x: NDArray | None
+    start_fun: float
+    solution_fun: float
+```
+
+### `run_benchmark`
+
+#### Current situation
+
+`run_benchmark` takes `benchmark_problems` (covered in the previous section),
+`optimize_options` and a few other arguments and returns a nested dictionary
+representing benchmark results.
+
+`optimize_options` can be a list of algorithm names, a dict with algorithm names as
+values or a nested dict of keyword arguments for `minimize`.
+
+**Things we want to keep**
+
+- Collecting benchmark results in a dictionary is good because it makes it easy to merge
+  results from multiple benchmark runs or filter results. A fixed field data structure
+  would not work here.
+
+**Problems**
+
+- `optimize_options` are super flexible but error prone and hard to write as there is no
+  autocomplete support
+- Each single benchmark result should not be represented as a dictionary
+
+#### Proposal
+
+We restrict the typo of `optimize_options` to
+`dict[str, Type[Algorithm] | Algorithm | OptimizeOptions]`. Here, `OptimizeOptions` will
+be a simple dataclass that we need for `estimate_ml` and `estimate_msm` anyways.
+
+Passing just lists of algorithm names is deprecated. Passing dicts as optimize options
+is also deprecated. Most use-cases will be covered by passing dictionaries of configured
+Algorithms as optimize options. Actually using the full power of passing
+`OptimizeOptions` will be rarely needed.
+
+The return type of `run_benchmark` will be `dict[tuple[str], BenchmarkResult]`
+
+`BenchmarkResult` is a dataclass with fields that mirror the keys of the current
+dictionary. It will roughly look as follows:
+
+```python
+@dataclass
+class BenchmarkResult:
+    params_history: list[NDArray]
+    fun_history: list[float]
+    time_history: list[float]
+    batches_history: list[int]
+    solution: OptimizeResult
+```
+
+## Estimation
+
+The changes to the estimation functions `estimate_ml` and `estimate_msm` will be
+minimal:
+
+- `lower_bounds` and `upper_bounds` are replaced by `bounds` (as in optimization)
+- `numdiff_options` and `optimize_options` become dataclasses
+- `logging` and `log_options` get aligned with our proposal for optimization
+
+In the long run we plan a general overhaul of `MSM` estimation that provides better
+access to currently internal objects such as the MSM objective function.
+
+(design-philosophy)=
+
+## Design Philosophy
```

Comment:
I think this should go upfront, i.e., after "Motivation and resources". *(makes sense you wrote it last, but for a reader it is better to have upfront)*

Reply:
Moved it to the front, thanks!